Avoid considering EmptyOperator in mini scheduler

eladkal opened the pull request:
EmptyOperator should not be executed on workers thus it should not be considered for the mini scheduler optimization. closes: https://github.com/apache/airflow/issues/29974


TODO:
- [x] add test

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

eladkal committed:
Avoid considering EmptyOperator in mini scheduler

EmptyOperator should not be executed on workers thus it should not be considered for the mini scheduler optimization.
closes: https://github.com/apache/airflow/issues/29974

eladkal committed:
add test

ephraimbuddy commented on airflow/models/taskinstance.py:
```diff
@@ -2651,7 +2651,11 @@ def schedule_downstream_tasks(self, session=None):
+                ti
+                for ti in info.schedulable_tis
+                if ti.task_id not in skippable_task_ids and not ti.task.inherits_from_empty_operator
+            ]
```
I have some doubts; `dagrun.schedule_tis` marks empty operators as successful when called by the scheduler, how does it still get to this point?

eladkal replied:
This is schedule_downstream_tasks for the mini-scheduler
It's being called in 
https://github.com/apache/airflow/blob/f9e9d23457cba5d3e18b5bdb7b65ecc63735b65b/airflow/jobs/local_task_job.py#L225
as part of `handle_task_exit` which is when the task is finished. Basically when task is finished the mini scheduler kicks in. We already have the dag loaded and we assume direct downstream tasks are next to be scheduled so if we can set them here it saves the need for scheduler to scan them in it's next loop.

eladkal committed:
add handling for cases when we do want to consider EmptyOperator in mini scheduler.
for example: EmptyOperator with on_execute_callback, on_success_callback, outlets

jedcunningham commented on tests/models/test_taskinstance.py:
```diff
@@ -3859,6 +3859,50 @@ def last_task():
+        forth_task = dag_run.get_task_instance(task_id="forth_task")
+        assert second_task.state != State.NONE
+        assert third_task.state == State.NONE
+        assert forth_task.state != State.NONE
```
```suggestion
        assert second_task.state == State.SCHEDULED
        assert third_task.state == State.NONE
        assert forth_task.state == State.SCHEDULED
```

nit, feel more natural to check what it is, not what it isn't?

eladkal committed:
change state
